libkbfs/rekey_fsm.go

```diff
@@ -0,0 +1,526 @@
+	return &rekeyStateIdle{fsm: fsm}
+}
+
+func (r *rekeyStateIdle) reactToEvent(event RekeyEvent) rekeyState {
+	switch event.eventType {
+	case rekeyRequestEvent:
+		return newRekeyStateScheduled(r.fsm,
+			event.request.delay, event.request.rekeyTask)
+	default:
+		return r
+	}
+}
+
+type rekeyStateScheduled struct {
+	fsm *rekeyFSM
+
+	timer    *time.Timer
+	deadline time.Time
+
+	task rekeyTask
+}
+
+func newRekeyStateScheduled(
+	fsm *rekeyFSM, delay time.Duration, task rekeyTask) *rekeyStateScheduled {
+	task.ctx.setLogger(fsm.log)
+	return &rekeyStateScheduled{
+		fsm: fsm,
+		timer: time.AfterFunc(delay, func() {
+			fsm.Event(newRekeyTimeupEvent())
+		}),
+		deadline: time.Now().Add(delay),
+		task:     task,
+	}
+}
+
+func (r *rekeyStateScheduled) reactToEvent(event RekeyEvent) rekeyState {
+	switch event.eventType {
+	case rekeyTimeupEvent:
+		return newRekeyStateStarted(r.fsm, r.task)
+	case rekeyRequestEvent:
+		task := r.task
+		task.promptPaper = task.promptPaper || event.request.promptPaper
+		if task.timeout == nil {
+			task.timeout = event.request.timeout
+		}
+		task.ttl = event.request.ttl
+		task.ctx.maybeReplaceContext(event.request.ctx.context())
+		if !r.deadline.After(time.Now().Add(event.request.delay)) {
+			r.fsm.log.CDebugf(task.ctx.context(), "Reusing existing timer")
+			r.task = task
+			return r
+		}
+		r.timer.Stop()
+		return newRekeyStateScheduled(r.fsm, event.request.delay, task)
+	case rekeyKickoffEventForTest:
+		r.timer.Reset(time.Millisecond)
+		return r
+	case rekeyCancelEventForTest:
+		r.timer.Stop()
+		return newRekeyStateIdle(r.fsm)
+	case rekeyShutdownEvent:
+		r.timer.Stop()
+		return r
+	default:
+		return r
+	}
+}
+
+type rekeyStateStarted struct {
+	fsm  *rekeyFSM
+	task rekeyTask
+}
+
+func newRekeyStateStarted(fsm *rekeyFSM, task rekeyTask) *rekeyStateStarted {
+	ctx := task.ctx.context()
+	var cancel context.CancelFunc
+	if task.timeout != nil {
+		ctx, cancel = context.WithTimeout(task.ctx.context(), *task.timeout)
+	}
+	go func() {
+		if cancel != nil {
+			defer cancel()
+		}
+		fsm.log.CDebugf(ctx, "Processing rekey for %s", fsm.fbo.folderBranch.Tlf)
+		var res RekeyResult
+		err := fsm.fbo.doMDWriteWithRetryUnlessCanceled(ctx,
+			func(lState *lockState) (err error) {
+				res, err = fsm.fbo.rekeyLocked(ctx, lState, task.promptPaper)
+				return err
+			})
+		fsm.log.CDebugf(ctx, "Rekey finished with res=%#+v, error=%v", res, err)
+		fsm.Event(newRekeyFinishedEvent(res, err))
+	}()
+	return &rekeyStateStarted{
+		fsm:  fsm,
+		task: task,
+	}
+}
+
+func (r *rekeyStateStarted) reactToEvent(event RekeyEvent) rekeyState {
+	switch event.eventType {
+	case rekeyFinishedEvent:
+		ttl := r.task.ttl - 1
+		r.fsm.log.CDebugf(r.task.ctx.context(),
+			"Rekey finished, ttl: %d -> %d", r.task.ttl, ttl)
+
+		if ttl <= 0 {
+			r.fsm.log.CDebugf(r.task.ctx.context(),
+				"Not scheduling new rekey because TTL expired")
+			return newRekeyStateIdle(r.fsm)
+		}
+
+		switch event.finished.err {
```

Comment:
I think you mentioned you'd be adding a explicit check for `context.DeadlineExceeded` (and maybe add a 10 min delay instead of 0), but this still seems to retry no matter what the error.